### PR TITLE
Enrich build trace with the culprits

### PR DESCRIPTION
**Pull request by v1v**
- [x] Make sure you are opening from a **topic/feature/bugfix branch** (right side) and not your main branch!
- [x] Ensure that the pull request title represents the desired changelog entry
- [x] Please describe what you did
- [x] Link to relevant issues in GitHub or Jira
- [ ] Link to relevant pull requests, esp. upstream and downstream changes
- [x] Ensure you have provided tests - that demonstrates feature works or fixes the issue

### What

Enrich the trace with the culprits of the existing build.

### Task

- [x] UTs
- [x] ITs

### Issues

Closes https://github.com/jenkinsci/opentelemetry-plugin/issues/224

### Questions

- [ ] Singular or plural attribute?
- [ ] `culprit` might have some negative context, what other synonym could we use?

### CODE REVIEW

File: src/main/java/io/jenkins/plugins/opentelemetry/job/MonitoringRunListener.java
Comment (mdelapenya) [inline src/main/java/io/jenkins/plugins/opentelemetry/job/MonitoringRunListener.java:141]: What does the `culpritId` hold reference for, the user email? Will it work outside Git world?

Comment (v1v) [reply src/main/java/io/jenkins/plugins/opentelemetry/job/MonitoringRunListener.java:141]: > What does the culpritId hold reference for, the user email? Will it work outside Git world?

It's provided by Jenkins core -> https://javadoc.jenkins.io/hudson/model/User.html 

This is parsed from https://javadoc.jenkins.io/jenkins/scm/RunWithSCM.html#calculateCulprits--

In other words, any plugin that implements https://github.com/jenkinsci/scm-api-plugin/blob/d01f419bc9dcc408e834cbc333b285b1c8058e69/docs/implementation.adoc#implementing-hudsonscmscm can produce that particular output.

> Is there a way to notify the jUnit extension about this culprits? Or simply using the same code as in here would be enough?

Once we merge then we will benefit from since these attributes are for the transaction. Meanwhile, I'ld say to use the same implementation.

File: src/test/java/io/jenkins/plugins/opentelemetry/BaseIntegrationTest.java
Comment (v1v) [inline src/test/java/io/jenkins/plugins/opentelemetry/BaseIntegrationTest.java:175]: When running the same job a few times, this tree always returns the first build.

File: src/test/java/io/jenkins/plugins/opentelemetry/JenkinsOtelPluginFreestyleIntegrationTest.java
Comment (v1v) [inline src/test/java/io/jenkins/plugins/opentelemetry/JenkinsOtelPluginFreestyleIntegrationTest.java:230]: Assertion for the build 3 that contains the culprit `bob`.

File: src/main/java/io/jenkins/plugins/opentelemetry/job/MonitoringRunListener.java
Comment (cyrille-leclerc) [inline src/main/java/io/jenkins/plugins/opentelemetry/job/MonitoringRunListener.java:143]: Can we just do something like:

```java
Set<User> culpritIds;
if (run instanceof WorkflowRun) {
   culpritIds = ((WorkflowRun) run).getCulprits();
} else if (run instanceof AbstractBuild) {
   culpritIds = ((AbstractBuild) run).getCulprits();
} else {
   culpritIds = null;
}
if (culpritIds != null) {
   rootSpanBuilder
      .setAttribute(JenkinsOtelSemanticAttributes.CI_PIPELINE_RUN_CULPRITS,
      culpritIds.stream().map(p -> p.getId()).collect(Collectors.toList()));
}
```

File: src/main/java/io/jenkins/plugins/opentelemetry/job/MonitoringRunListener.java
Comment (cyrille-leclerc) [inline src/main/java/io/jenkins/plugins/opentelemetry/job/MonitoringRunListener.java:142]: Why not default to null

**Review by cyrille-leclerc — CHANGES_REQUESTED** [review]
looks great

File: src/main/java/io/jenkins/plugins/opentelemetry/job/MonitoringRunListener.java
Comment (v1v) [reply src/main/java/io/jenkins/plugins/opentelemetry/job/MonitoringRunListener.java:142]: Do you mean to avoid sending the attribute if no matches?

File: README.md
Comment (v1v) [inline README.md:132]: Singular or plural?

File: src/main/java/io/jenkins/plugins/opentelemetry/semconv/JenkinsOtelSemanticAttributes.java
Comment (v1v) [inline src/main/java/io/jenkins/plugins/opentelemetry/semconv/JenkinsOtelSemanticAttributes.java:33]: Should we use a different term rather than culprits? Thinking about the standardisation. @mdelapenya pointed to use `committers` IIRC.

Though, we can change it in the future anyway, since we will probably revisit the attribute names

**Review by kuisathaverat — APPROVED** [review]

**Comment by cyrille-leclerc** [comment]
@v1v Are all you comments answered?

**Comment by v1v** [comment]
There are some open questions regarding:
- plural or singular in the attribute name
- Name of the attribute

There are no blockers but if we can agree then we can reduce the future work around the standards in the attributes

**Comment by v1v** [comment]
any feedback regarding my comment in https://github.com/jenkinsci/opentelemetry-plugin/pull/223#issuecomment-986273155 ? 🙏

File: README.md
Comment (cyrille-leclerc) [reply README.md:132]: I like plural for string arrays.
Shouldn't this attribute be documented as a `String[]` / String array rather than as a `String`?

File: src/main/java/io/jenkins/plugins/opentelemetry/semconv/JenkinsOtelSemanticAttributes.java
Comment (cyrille-leclerc) [reply src/main/java/io/jenkins/plugins/opentelemetry/semconv/JenkinsOtelSemanticAttributes.java:33]: if it's just listing the committers of the changes of this build, I like the idea to call the attribute `CI_PIPELINE_RUN_COMMITTERS`

File: README.md
Comment (v1v) [reply README.md:132]: Uhm, I'm not sure now, since the `cause` uses the same approach, a list of causes, but the attribute name is singular.

Comment (cyrille-leclerc) [reply README.md:132]: Good catch. I feel it's `ci.pipeline.run.cause` that should be renamed `ci.pipeline.run.causes` as it's a string array.

**Pull request merged**
